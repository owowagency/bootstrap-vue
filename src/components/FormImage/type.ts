const dataTypes = ['file', 'base64'] as const;

type DataType = typeof dataTypes[number];

export {dataTypes, DataType};
