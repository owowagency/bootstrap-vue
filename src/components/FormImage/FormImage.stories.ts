import FormImage from '.';
import {ref} from 'vue';

export default {
    title: 'Forms/Form image',
    component: FormImage,
    argTypes: {},
};

const template = (args) => ({
    components: {FormImage},
    setup: () => ({args, value: ref<string>()}),
    template: '<FormImage v-model="value" :preview="preview" v-bind="args" />',
});

export const Default = template.bind({});
