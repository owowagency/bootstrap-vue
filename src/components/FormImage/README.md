# FormImage

```vue
<FormImage v-model="b64" type="base64" />
```

## Imported Props

| Prop name | Description                   | Type   | Values | Default |
| --------- | ----------------------------- | ------ | ------ | ------- |
| id        | The id used for the component | string | -      | `uuid`  |

## Props

| Prop name  | Description                | Type         | Values             | Default   |
| ---------- | -------------------------- | ------------ | ------------------ | --------- |
| dataType   | The data type of the model | DataType     | `'file'\|'base64'` | 'base64'  |
| modelValue | The model value            | File\|string | -                  | undefined |
| preview    | The preview                | string       | -                  | undefined |

## Slots

| Name        | Description              | Bindings |
| ----------- | ------------------------ | -------- |
| placeholder | Displays the placeholder |          |

## Events

| Event name        | Properties                        | Description                           |
| ----------------- | --------------------------------- | ------------------------------------- |
| update:modelValue | **value** `union` - The new value | Fired when the model value is updated |
