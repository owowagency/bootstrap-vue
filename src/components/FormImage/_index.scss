.form-image {
    position: relative;
    display: block;
    @include border-radius($form-image-border-radius);
    border: $form-image-border-width solid $form-image-border-color;
    width: $form-image-width;
    height: $form-image-height;
    cursor: pointer;
    background-color: $form-image-background-color;
    background-image: escape-svg($form-image-upload-icon);
    background-position: center;
    background-repeat: no-repeat;
    background-size: $form-image-upload-icon-size;

    .form-image-btn-edit {
        position: absolute;
        top: $form-image-btn-edit-top;
        right: $form-image-btn-edit-right;
    }
}

.form-image-btn-edit {
    @include border-radius($form-image-btn-edit-border-radius);
    width: $form-image-btn-edit-width;
    height: $form-image-btn-edit-height;
    background-color: $form-image-btn-edit-bg-color;
    background-image: escape-svg($form-image-btn-edit-icon);
    background-position: center;
    background-repeat: no-repeat;
    background-size: $form-image-btn-edit-icon-size;
}
