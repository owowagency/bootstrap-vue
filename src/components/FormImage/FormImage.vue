<template>
    <label
        class="form-image"
        :class="{'form-image-has-preview': !!preview}"
        :for="id"
        :style="style"
        @drop="onDrop"
        @dragenter.prevent
        @dragover.prevent
    >
        <input
            :id="id"
            ref="input"
            accept="image/*"
            hidden
            type="file"
            @change="onChange"
        >

        <div
            v-if="preview"
            class="form-image-btn-edit"
        />

        <!-- @slot Displays the placeholder -->
        <slot
            v-else
            name="placeholder"
        />
    </label>
</template>

<script lang="ts">
import {PropType, computed, ref, watch} from 'vue';
import {DataType} from '@/components/FormImage/type';
import {idProps} from '@/composables/useId';
</script>

<script lang="ts" setup>
const props = defineProps({
    /**
     * The data type of the model
     * @values 'file'|'base64'
     */
    dataType: {
        type: String as PropType<DataType>,
        default: 'base64',
    },
    ...idProps,
    /**
     * The model value
     */
    modelValue: {
        type: [Object, String] as PropType<File|string>,
        default: undefined,
    },
    /**
     * The preview
     */
    preview: {
        type: String,
        default: undefined,
    },
});

const emit = defineEmits<{
    /**
     * Fired when the model value is updated
     * @param {string|File} value The new value
     */
    (e: 'update:modelValue', value: string|File): void,
}>();

// Rollup cannot handle HTMLInputElement as ref, cannot
const input = ref();

const file = ref<File|undefined>();

const readerResult = ref<string|undefined>();

const preview = computed(() => readerResult.value || props.preview);

const style = computed(() => {
    if (!preview.value) {
        return {};
    }

    return {
        'background-color': 'transparent',
        'background-image': `url(${preview.value})`,
        'background-size': 'cover',
    };
});

const value = computed<string|File|undefined>({
    get: () => props.modelValue,
    set: (v?: string|File) => emit('update:modelValue', v),
});

watch(file, f => readFile(f));

const onChange = (e: {target: HTMLInputElement}) => {
    const newFile = e.target.files.item(0);

    if (newFile?.type.match('image.*')) {
        file.value = newFile;

        return true;
    }

    return false;
};

const onDrop = (e: DragEvent) => {
    e.preventDefault();

    input.value.files = e.dataTransfer.files;

    onChange({target: input.value});
};

const readFile = (file: File) => {
    if (!file) {
        return value.value = undefined;
    }

    if (props.dataType === 'file') {
        value.value = file;
    }

    const reader = new FileReader();

    reader.onload = () => {
        readerResult.value = reader.result as string;

        if (props.dataType === 'base64') {
            value.value = reader.result as string;
        }
    };

    reader.readAsDataURL(file);

    return reader;
};
</script>

<docs>
```vue
<FormImage
    v-model="b64"
    type="base64"
/>
```

## Imported Props

| Prop name | Description                   | Type   | Values | Default |
| --------- | ----------------------------- | ------ | ------ | ------- |
| id        | The id used for the component | string | -      | `uuid`  |
</docs>
