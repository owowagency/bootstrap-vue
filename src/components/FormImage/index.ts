import FormImage from './FormImage.vue';
export default FormImage;
export * from './type';
