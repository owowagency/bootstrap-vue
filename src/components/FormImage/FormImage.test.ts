import FormImage from '.';
import {nextTick} from 'vue';
import {shallowMount} from '@vue/test-utils';

describe('template', () => {
    componentRenderTest(FormImage, {
        // Force an id to prevent snapshots from failing because of uuid.
        props: {id: 'dont-change-please'},
    });

    componentWrapperClassTest(
        FormImage,
        {preview: 'some-preview'},
        'form-image-has-preview',
    );

    it('renders btn edit when has preview', () => {
        const wrapper = shallowMount(FormImage, {props: {preview: 'some-preview'}});

        expect(wrapper.find('.form-image-btn-edit').exists()).toBe(true);
    });

    it('adds class when has preview', () => {
        const wrapper = shallowMount(FormImage, {props: {preview: 'some-preview'}});

        expect(wrapper.find('.form-image-btn-edit').exists()).toBe(true);
    });

    // TODO: Unable to test drop, dragenter, dragover properly. It is currently
    // not possible to create a FileList instance used during the assigment of
    // the files to a file input element. It is currently not possible to mock
    // methods defined in setup.
});

describe('file', () => {
    describe('watch', () => {
        it('sets value to undefined when not thruthful', async() => {
            const wrapper = shallowMount(FormImage, {props: {modelValue: 'some-file'}});

            // First change to something that is not undefined but still falsly
            // to prevent the watcher to read the value as a file.
            wrapper.vm.file = false;

            await nextTick();

            wrapper.vm.file = undefined;

            // Cannot test value to be exactly undefined since it is a computed
            // property that will emit when set.
            expect(wrapper.emitted('update:modelValue')[0]).toEqual([undefined]);
        });

        // More detailed tests are done in readFile tests.
    });
});

describe('style', () => {
    it('is empty when there is no preview', () => {
        const wrapper = shallowMount(FormImage);

        expect(wrapper.vm.style).toEqual({});
    });

    it('has background styling when there is a preview', () => {
        const wrapper = shallowMount(FormImage);

        wrapper.vm.preview = 'some-preview';

        expect(wrapper.vm.style).toEqual({
            'background-color': 'transparent',
            'background-image': 'url(some-preview)',
            'background-size': 'cover',
        });
    });
});

describe('value', () => {
    it('is equal to modelValue', () => {
        const wrapper = shallowMount(
            FormImage,
            {props: {modelValue: 'some-file'}},
        );

        expect(wrapper.vm.value).toBe('some-file');
    });

    it('emits update:modelValue when set', () => {
        const wrapper = shallowMount(FormImage);

        wrapper.vm.value = 'some-value';

        expect(wrapper.emitted('update:modelValue')[0]).toEqual(['some-value']);
    });
});

describe('onChange', () => {
    it('skips empty files', () => {
        const wrapper = shallowMount(FormImage);

        const event = {target: {files: {item: () => undefined}}};

        expect(wrapper.vm.onChange(event)).toBe(false);
    });

    it('skips empty non image files', () => {
        const wrapper = shallowMount(FormImage);

        const file = new File([], 'test.json', {type: 'application/json'});

        const event = {target: {files: {item: () => file}}};

        expect(wrapper.vm.onChange(event)).toBe(false);
    });

    it('sets file value when image', () => {
        const wrapper = shallowMount(FormImage);

        const file = new File([], 'test.png', {type: 'image/png'});

        const event = {target: {files: {item: () => file}}};

        expect(wrapper.vm.onChange(event)).toBe(true);

        expect(wrapper.vm.file).toBe(file);
    });
});

describe('onDrop', () => {
    it('calls onChange with files', () => {
        const wrapper = shallowMount(FormImage);

        const event = {
            preventDefault: jest.fn(),
            dataTransfer: {
                files: {
                    // Mock this method to test that `onChange` is being
                    // called.
                    item: jest.fn()
                        .mockImplementation(() => undefined),
                },
            },
        };

        // Overwrite the input ref to prevent type "files should be of type
        // FileList" error.
        wrapper.vm.input = {files: {}};

        wrapper.vm.onDrop(event);

        expect(event.preventDefault).toBeCalled();

        expect(event.dataTransfer.files.item).toBeCalledWith(0);
    });
});

describe('readFile', () => {
    // The truthful test is done in watcher test of file to test if this method
    // is called.

    it('handles when dataType is file', async() => {
        const readAsDataURLSpy = jest.spyOn(FileReader.prototype, 'readAsDataURL');

        const wrapper = shallowMount(FormImage, {props: {dataType: 'file'}});

        const file = new File([], 'test.png');

        const reader = wrapper.vm.readFile(file);

        // Cannot test value to be exactly undefined since it is a computed
        // property that will emit when set.
        expect(wrapper.emitted('update:modelValue')[0]).toEqual([file]);

        expect(readAsDataURLSpy).toBeCalledWith(file);

        Object.defineProperty(reader, 'result', {value: 'some-preview'});

        reader.onload();

        expect(wrapper.vm.preview).toBe('some-preview');
    });

    it('handles when dataType is base64', async() => {
        const readAsDataURLSpy = jest.spyOn(FileReader.prototype, 'readAsDataURL');

        const wrapper = shallowMount(FormImage, {props: {dataType: 'base64'}});

        const file = new File([], 'test.png');

        const reader = wrapper.vm.readFile(file);

        expect(readAsDataURLSpy).toBeCalledWith(file);

        Object.defineProperty(reader, 'result', {value: 'some-preview'});

        reader.onload();

        expect(wrapper.vm.preview).toBe('some-preview');

        // Cannot test value to be exactly undefined since it is a computed
        // property that will emit when set.
        expect(wrapper.emitted('update:modelValue')[0]).toEqual(['some-preview']);
    });
});
