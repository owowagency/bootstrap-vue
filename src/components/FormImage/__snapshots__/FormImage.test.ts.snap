// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`template > renders default 1`] = `"<label class=\\"form-image\\" for=\\"dont-change-please\\"><input id=\\"dont-change-please\\" accept=\\"image/*\\" hidden=\\"\\" type=\\"file\\"><!-- @slot Displays the placeholder --></label>"`;
